abilities/garage-door-switch.js

```diff
@@ -0,0 +1,212 @@
+    }
+
+    get state() {
+      return this.device[this._stateProperty] || 0
+    }
+
+    get isSwitchedOn() {
+      return this.device[this._switchProperty] || false
+    }
+
+    get currentState() {
+      const CDS = Characteristic.CurrentDoorState
+
+      if (this.state > 0) {
+        return CDS.CLOSED
+      }
+
+      return CDS.OPEN
+    }
+
+    get targetState() {
+      if (this._targetState !== null) {
+        return this._targetState
+      }
+
+      const CDS = Characteristic.CurrentDoorState
+      const TDS = Characteristic.TargetDoorState
+      const cs = this.currentState
+      return cs === CDS.OPEN || cs === CDS.OPENING ? TDS.OPEN : TDS.CLOSED
+    }
+
+    _setupPlatformAccessory() {
+      const CDS = Characteristic.CurrentDoorState
+      const TDS = Characteristic.TargetDoorState
+      super._setupPlatformAccessory()
+
+      // This is the initial setup of the garage door
+      this.platformAccessory.addService(
+        new Service.GarageDoorOpener()
+          .setCharacteristic(CDS, this.currentState)
+          .setCharacteristic(TDS, TDS.CLOSED)
+          .setCharacteristic(Characteristic.ObstructionDetected, false)
+      )
+    }
+
+    _setupEventHandlers() {
+      super._setupEventHandlers()
+
+      // This is the handler to catch HomeKit events
+      this.platformAccessory
+        .getService(Service.GarageDoorOpener)
+        .getCharacteristic(Characteristic.TargetDoorState)
+        .on('set', this._targetDoorStateSetHandler.bind(this))
+
+      this.platformAccessory
+        .getService(Service.GarageDoorOpener)
+        .getCharacteristic(Characteristic.CurrentDoorState)
+        .on('get', this.getState.bind(this))
+
+      this.platformAccessory
+        .getService(Service.GarageDoorOpener)
+        .getCharacteristic(Characteristic.TargetDoorState)
+        .on('get', this.getState.bind(this))
+
+      // This is the handler to catch device events
+      // This one is always correct!
+      this.device
+        .on(
+          'change:' + this._stateProperty,
+          this._stateChangeHandler,
+          this
+        )
+    }
+
+    /**
+     * Handles changes from HomeKit to the TargetDoorState characteristic.
+     */
+    async _targetDoorStateSetHandler(newValue, callback, context) {
+      const d = this.device
+
+      // If the context is shelly then this is an internal update
+      // to ensure that homekit is in sync with the current status
+      // If not, we really trigger the switch
+      if (context === 'shelly') {
+        callback()
+        return
+      }
+
+      this._targetState = newValue
+
+      this.log.debug(
+        'Target homekit state is',
+        newValue
+      )
+
+      this.log.debug(
+        'Setting',
+        this._switchProperty,
+        'of device',
+        d.type,
+        d.id,
+        'to',
+        true
+      )
+
+      try {
+        await this._setSwitch(true)
+        callback()
+        this.updateGarageDoorState()
```

Comment:
What's the reason for calling `updateGarageDoorState()` here?

If the garage door is closed and the switch is turned on to open the door, the reed sensor will probably still be closed (e.g. `currentState === CLOSED`), and the current state will be set to closing, not opening. Briefly afterwards though, the reed sensor will be set to opened (e.g. `currentState === OPEN`) and the current state will be set to opening (which is correct). I suppose this still works but there might be some flickering in the UI? Am I missing something?